File: Bai01/Bai001_LTDD-main/app/src/main/java/com/example/baitap1/MainActivity.java
package com.example.baitap1;

import android.os.Bundle;
import android.util.Log;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class MainActivity extends AppCompatActivity {
    private EditText inputEditText;
    private Button processButton;
    private TextView resultTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        // thiết lập giao diện
        setContentView(R.layout.activity_main);

        //ánh xạ các view
        inputEditText = findViewById(R.id.inputEditText);
        processButton = findViewById(R.id.processButton);
        resultTextView = findViewById(R.id.resultTextView);

        // xử lí khi người dùng nhấn nút
        processButton.setOnClickListener(v -> processString());
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        processArray();

    }
    private void processArray() {
        Random random = new Random();
        int n = 10; // Số lượng phần tử ngẫu nhiên
        ArrayList<Integer> numbers = new ArrayList<>();

        // Sinh ra các số ngẫu nhiên
        for (int i = 0; i < n; i++) {
            numbers.add(random.nextInt(100)); // Sinh số ngẫu nhiên từ 0 đến 99
        }

        // Tạo ArrayList để lưu trữ các số lẻ và số chẵn
        ArrayList<Integer> oddNumbers = new ArrayList<>();
        ArrayList<Integer> evenNumbers = new ArrayList<>();

        // Phân loại các số lẻ và số chẵn
        for (int num : numbers) {
            if (num % 2 == 0) {
                evenNumbers.add(num);
            } else {
                oddNumbers.add(num);
            }
        }

        // In kết quả ra console (Logcat)
        Log.d("EvenNumbers", evenNumbers.toString());
        Log.d("OddNumbers", oddNumbers.toString());
    }
    private void processString() {
        // Lấy chuỗi từ EditText
        String inputString = inputEditText.getText().toString();

        // Đảo ngược chuỗi và chuyển thành chữ hoa
        String reversedString = reverseString(inputString);
        String upperCaseString = reversedString.toUpperCase();

        // Hiển thị kết quả lên TextView
        resultTextView.setText(upperCaseString);

        // Hiển thị kết quả qua Toast
        Toast.makeText(this, upperCaseString, Toast.LENGTH_SHORT).show();

        // In ra Logcat
        Log.d("ProcessedString", upperCaseString);
    }

    private String reverseString(String input) {
        String[] words = input.split(" "); // Tách chuỗi thành các từ
        StringBuilder reversedString = new StringBuilder();

        // Đảo ngược các từ trong chuỗi
        for (int i = words.length - 1; i >= 0; i--) {
            reversedString.append(words[i]).append(" ");
        }

        // Trả về chuỗi đảo ngược
        return reversedString.toString().trim();
    }

}

File: Bai05/LAB1/SQLite/app/src/main/java/com/example/sqlite_course/MainActivity.kt
package com.example.sqlite_course
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class MainActivity : AppCompatActivity() {
    private lateinit var databaseHelper: DatabaseHelper
    private lateinit var editTextCourseName: EditText
    private lateinit var editTextCourseDescription: EditText
    private lateinit var buttonAddCourse: Button
    private lateinit var recyclerViewCourses: RecyclerView
    private lateinit var courseAdapter: CourseAdapter
    private var courses: MutableList<Course> = mutableListOf()
    private var selectedCourseId: Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
// DatabaseHelper
        databaseHelper = DatabaseHelper(this)
// Views
        editTextCourseName =
            findViewById(R.id.editTextCourseName)
        editTextCourseDescription =
            findViewById(R.id.editTextCourseDescription)
        buttonAddCourse = findViewById(R.id.buttonAddCourse)
        recyclerViewCourses =
            findViewById(R.id.recyclerViewCourses)
// RecyclerView
        recyclerViewCourses.layoutManager =
            LinearLayoutManager(this)
        courseAdapter = CourseAdapter(this, courses, { course -> updateCourse(course) }, { id -> deleteCourse(id) })
        recyclerViewCourses.adapter = courseAdapter
// Add Course
        buttonAddCourse.setOnClickListener {
            addOrUpdateCourse() }
        loadCourses()
    }
    // Add or Update
    private fun addOrUpdateCourse() {
        val name = editTextCourseName.text.toString()
        val description =
            editTextCourseDescription.text.toString()
        if (name.isNotEmpty() && description.isNotEmpty()) {
// id == null > Add
            if (selectedCourseId == null) {
// Add new course
                val course = Course(0, name, description) // IDwill be auto-incremented
                if (databaseHelper.addCourse(course)) {
                    loadCourses()
                    clearInputFields()
                }
            } else {
// id != null > Update existing course
                val course = Course(selectedCourseId!!, name,
                    description)
                if (databaseHelper.updateCourse(course)) {
                    loadCourses() // Refresh the list after update
                    clearInputFields()
                    selectedCourseId = null
                    buttonAddCourse.text = "Add Course" // Reset button text
                    buttonAddCourse.visibility = View.VISIBLE
// Show the button again
                }
            }
        }
    }
    // Update data
    private fun loadCourses() {
        courses.clear()
        courses.addAll(databaseHelper.getCourses())
        courseAdapter.updateCourses(courses)
    }
    // Clear views
    private fun clearInputFields() {
        editTextCourseName.text.clear()
        editTextCourseDescription.text.clear()
    }
    // Click buton Update
    private fun updateCourse(course: Course) {
        editTextCourseName.setText(course.name)
        editTextCourseDescription.setText(course.description)
        selectedCourseId = course.id
        buttonAddCourse.text = "Update Course" // Change button text to indicate updating
        buttonAddCourse.visibility = View.VISIBLE // Show the button during editing
    }
    // Click buton Delete
    private fun deleteCourse(id: Int) {
        if (databaseHelper.deleteCourse(id)) {
            loadCourses()
        }
    }
}